### ISSUE TITLE
Tests do not wait for execution BeforeGroups method

### ISSUE THREAD

**Issue by choojoykin** [cbeust/testng#1859]
### TestNG Version
6.14.3

The tests with groups do not wait for the completion of the method in BeforeGroups by this groups. When first test run before groups method, other tests think that it is completed and run their test method. When before groups method has a timeout this causes problems in the tests that need data from it

### Expected behavior
In ver 6.9.10 all works fine. Expected - all tests in groups wait for complete BeforeGroups method

### Actual behavior
Tests do not wait for execution BeforeGroups method

### Is the issue reproductible on runner?

- [ ] Shell
- [x ] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [x ] IntelliJ
- [ ] NetBeans

### Test case sample
In this example i have NPE on random test when i want get data from list, cause i have sleep (e.g. data prepare or anything else)
output => 
```
setup start
second test

java.lang.NullPointerException
	at debug.Debug.getStringFromList(Debug.java:56)
	at debug.Debug.testTwo(Debug.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:583)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:719)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:989)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:109)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

setUp done
first test
first A
```
```java
public class Debug {

 List<String> dataList;

    @BeforeGroups("group")
    public void setup() {
        System.out.println("setup");
                try {
            Thread.sleep(5000); //without sleep this bug doesnt reproduce
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        dataList = new ArrayList<>(Arrays.asList("A", "B")); //just an example
        System.out.println("setUp done");
    }

    @Test(groups = "group")
    public void testOne() {
        System.out.println("first test");
        String first = getStringFromList();
        System.out.println("first " + first);
    }

    @Test(groups = "group")
    public void testTwo() {
        System.out.println("second test");
        String second = getStringFromList();
        System.out.println("second " + second);
    }

    private String getStringFromList() {
        return dataList.stream().findFirst().orElse("C"); //just an example
    }
}
```

**Comment by krmahadevan** [comment]
Attempted reproducing this issue using `TestNG 7.0.0-beta1` with the above provided sample class and below sample xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="1859_suite" parallel="false" verbose="2">
  <test name="1859_test">
    <groups>
      <run>
        <include name="group"/>
      </run>
    </groups>
    <classes>
      <class name="com.rationaleemotions.github.issue1859.Debug"/>
    </classes>
  </test>
</suite>
```

Unable to recreate the problem. Closing this issue. Please comment if this is still a problem with `7.0.0-beta1` along with a suitable sample that can be used to recreate the problem.

**Comment by krshubham** [comment]
if you do parallel="methods", and thread-count="4" you will be able to reproduce it.
I am having this problem in my code also.

**Comment by choojoykin** [comment]
why it cant be solved in current last stable version? i didnt want use beta version in my production code. it still reproduce

**Comment by krmahadevan** [comment]
How do you expect this to be fixed in the current version without a code change ? If u aren't comfortable with using a beta version pls wait till 7.0.0 gets released

**Comment by krshubham** [comment]
@krmahadevan , is there an estimated time, around which 7.0.0 will get released?

**Comment by krmahadevan** [comment]
We are currently trying to resolve some issues with respect to publishing beta builds. The release may take sometime because we have tagged some issues which we want to fix before 7.0.0 gets released

**Comment by cerbule** [comment]
maybe is to late since is 2022 :), but always put after the annotation the always true parameter like this: @BeforeGroups(alwaysRun = true)
